File: library/src/main/java/bz/tsung/android/objectify/ObjectPreferenceLoader.java
package bz.tsung.android.objectify;

import android.content.Context;

import com.google.gson.Gson;

import java.lang.reflect.Type;

/**
 * Created by tsung on 3/18/14.
 */
public class ObjectPreferenceLoader {
    StringPreferenceLoader stringPreferenceLoader;
    Type clazz;
    Context context;
    Gson gson;

    public <T> ObjectPreferenceLoader(Context context, final String key, Type clazz) {
        this(context, key, clazz, new Gson());
    }

    public <T> ObjectPreferenceLoader(Context context, final String key, Type clazz, Gson gson) {
        this.context = context;
        this.clazz = clazz;
        stringPreferenceLoader = new StringPreferenceLoader(context, key);
        this.gson = gson;
    }

    public <T> ObjectPreferenceLoader(Context context, final String key, Class<T> clazz) {
        this(context, key, clazz, new Gson());
    }

    public <T> ObjectPreferenceLoader(Context context, final String key, Class<T> clazz, Gson gson) {
        this.context = context;
        this.clazz = clazz;
        stringPreferenceLoader = new StringPreferenceLoader(context, key);
        this.gson = gson;
    }

    public <T> T get() {
        try {
            String value = stringPreferenceLoader.load();
            return gson.fromJson(value, clazz);
        } catch (Exception e) {
            return null;
        }
    }

    public <T> void set(T value) {
        if ((clazz instanceof Class && ((Class)clazz).isInstance(value)) || clazz instanceof Type) {
            stringPreferenceLoader.save(gson.toJson(value));
        } else {
            throw new ShouldSaveSameTypeValueException();
        }
    }

    @Deprecated
    public <T> T load() throws NoSuchPreferenceFoundException {
        String value = stringPreferenceLoader.load();
        try {
            return gson.fromJson(value, clazz);
        } catch (Exception e) {
            throw new NoSuchPreferenceFoundException();
        }
    }

    @Deprecated
    public <T> void save(T value) {
        set(value);
    }

    public void remove() {
        stringPreferenceLoader.remove();
    }

    public static void clear(Context context) {
        StringPreferenceLoader.clear(context);
    }
}

File: library/src/main/java/bz/tsung/android/objectify/ObjectPreferenceLoader.kt
package bz.tsung.android.objectify

import android.content.Context
import com.google.gson.Gson
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import java.lang.reflect.Type

/**
 * Created by tsung on 3/18/14.
 */
class ObjectPreferenceLoader {
    var stringPreferenceLoader: StringPreferenceLoader
    var clazz: Type
    var context: Context
    var gson: Gson

    @JvmOverloads
    constructor(context: Context, key: String, clazz: Type, gson: Gson = Gson()) {
        this.context = context
        this.clazz = clazz
        stringPreferenceLoader = StringPreferenceLoader(context, key)
        this.gson = gson
    }

    fun <T> get(): T? {
        return try {
            stringPreferenceLoader.get()?.let {
                gson.fromJson<T>(it, clazz)
            }
        } catch (e: Throwable) {
            null
        }
    }

    fun <T> set(value: T) {
        if (clazz is Class<*> && (clazz as Class<*>).isInstance(value) || clazz is Type) {
            stringPreferenceLoader.set(gson.toJson(value))
        } else {
            throw ShouldSaveSameTypeValueException()
        }
    }

    fun <T> flow(): Flow<T?> {
        return stringPreferenceLoader.flow().map {
            it?.let { gson.fromJson<T>(it, clazz) }
        }
    }

    suspend fun <T> setSuspend(value: T) {
        if (clazz is Class<*> && (clazz as Class<*>).isInstance(value) || clazz is Type) {
            stringPreferenceLoader.setSuspend(gson.toJson(value))
        } else {
            throw ShouldSaveSameTypeValueException()
        }
    }

    fun remove() {
        stringPreferenceLoader.remove()
    }

    companion object {
        fun clear(context: Context?) {
            PreferenceLoader.clear(context)
        }
    }
}
